c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -223,7 +234,331 @@ struct Block {
+    TORCH_INTERNAL_ASSERT(ptr() + begin * segment_size_ == range.ptr);
+    if (begin == end) {
+      return rangeFromHandles(begin, end);
+    }
+    while (end > handles_.size()) {
+      handles_.push_back(c10::nullopt);
+    }
+    for (auto i : c10::irange(begin, end)) {
+      TORCH_INTERNAL_ASSERT(!handles_.at(i));
+      CUmemGenericAllocationHandle handle;
+      CUmemAllocationProp prop = {};
+      prop.type = CU_MEM_ALLOCATION_TYPE_PINNED;
+      prop.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+      prop.location.id = device_;
+      auto status = cuMemCreate(&handle, segment_size_, &prop, 0);
+      if (status == CUDA_ERROR_OUT_OF_MEMORY) {
+        for (auto j : c10::irange(begin, i)) {
+          auto handle = handles_.at(j).value();
+          handles_.at(j) = c10::nullopt;
+          C10_CUDA_DRIVER_CHECK(cuMemRelease(handle));
+        }
+        trimHandles();
+        return rangeFromHandles(begin, begin);
+      }
+      C10_CUDA_DRIVER_CHECK(status);
+      handles_.at(i) = handle;
+    }
+    for (auto i : c10::irange(begin, end)) {
+      C10_CUDA_DRIVER_CHECK(cuMemMap(
+          ptr_ + i * segment_size_,
+          segment_size_,
+          0,
+          handles_.at(i).value(),
+          0ULL));
+    }
+
+    setAccess(device_, begin, end);
+    for (auto p : peers_) {
+      setAccess(p, begin, end);
+    }
+    return rangeFromHandles(begin, end);
+  }
+
+  // unmaps all the completely empty segment_size_ segments between
+  // [begin, begin + size), returns the offset where the range begin,
+  // and the actual size unmapped (multiple of segment_size_)
+  SegmentRange unmap(SegmentRange range) {
+    auto begin = segmentRight(range.ptr);
+    auto end = segmentLeft(range.ptr + range.size);
+    if (begin >= end) {
+      return SegmentRange{range.ptr, 0};
+    }
+    unmapHandles(begin, end);
+    return rangeFromHandles(begin, end);
+  }
+
+  char* ptr() const {
+    return (char*)ptr_;
+  }
+  size_t size() const {
+    return max_handles_ * segment_size_;
+  }
+
+  void addPeer(int device) {
+    peers_.push_back(device);
+    forEachAllocatedRange(
+        [&](size_t begin, size_t end) { setAccess(device, begin, end); });
+  }
+
+  ~ExpandableSegment() {
+    forEachAllocatedRange(
+        [&](size_t begin, size_t end) { unmapHandles(begin, end); });
+    C10_CUDA_DRIVER_CHECK(cuMemAddressFree(ptr_, segment_size_ * max_handles_));
+  }
+
+ private:
+  void setAccess(int device, size_t begin, size_t end) {
+    CUmemAccessDesc desc;
+    desc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
+    desc.location.id = device;
+    desc.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+    C10_CUDA_DRIVER_CHECK(cuMemSetAccess(
+        ptr_ + begin * segment_size_, (end - begin) * segment_size_, &desc, 1));
+  }
+
+  void unmapHandles(size_t begin, size_t end) {
+    // note: unlike cudaFree, MemUnmap and MemRelease do
+    // not appear to synchronize in all cases, so we have to wait for the
```

Comment:
Docs say it doesn't guarantee synchronization:

<img width="624" alt="image" src="https://user-images.githubusercontent.com/11477974/231852867-928b95b4-e64a-41ed-b9ee-6ddbf7cc0362.png">